[Flang]: Lowering reference to functions that return a procedure pointer

DanielCChen opened the pull request:
This PR adds lowering the reference to a function that returns a procedure pointer. It also fixed intrinsic ASSOCIATED to take such argument.

DanielCChen committed:
[Flang]: Lowering reference to functions that return a procedure pointer.

DanielCChen committed:
[Flang]: Lowering reference to functions that return a procedure pointer.

DanielCChen committed:
[Flang]: Lowering reference to functions that return a procedure pointer.

DanielCChen committed:
[Flang]: Lowering reference to functions that return a procedure pointer.

DanielCChen committed:
[Flang]: Lowering reference to functions that return a procedure pointer.

DanielCChen committed:
[Flang]: Fix LIT test failures that resultType could be null.

DanielCChen committed:
[Flang] Add LIT test.

jeanPerier commented on flang/lib/Lower/ConvertCall.cpp:
```diff
@@ -156,7 +156,10 @@ fir::ExtendedValue Fortran::lower::genCallOpAndResult(
+
+  // Is procedure pointer functin result.
+  bool isProcedurePointer = resultType && resultType->isa<fir::BoxProcType>();
+  if (!isProcedurePointer && caller.mustMapInterfaceSymbols()) {
```
I think I would prefer `caller.mustMapInterfaceSymbols` and `caller.callerAllocateResult()` to return false in the Procedure pointer result case instead of overriding that here.

jeanPerier commented on flang/lib/Lower/ConvertExprToHLFIR.cpp:
```diff
@@ -1425,9 +1425,14 @@ class HlfirBuilder {
-        "lowering function references that return procedure pointers to HLFIR");
+    fir::FirOpBuilder &builder = getBuilder();
+    Fortran::evaluate::ProcedureDesignator proc{expr.proc()};
+    auto procTy{Fortran::lower::translateSignature(proc, getConverter())};
+    mlir::Type resType = fir::BoxProcType::get(builder.getContext(), procTy);
```
I am not convinced that resType is correct here.

I would expect resType to be something like `procTy.getResult(0)``, I am wondering if the current code isn't creating a `fir.boxproc<(ArgsX) -> (fir.boxproc((R) -> (ArgsY, ...)))>` instead of `fir.boxproc((R) -> (ArgsY, ...))`.

I think this makes little functional difference currently given resType is not used to generate IR (just to check if it is a fir::BoxProcType, which is true in both case), but could you check that recType is the expected `fir.boxproc((R) -> (ArgsY, ...))` here to avoid future bugs?

DanielCChen replied:
@jeanPerier Thanks for the review! Sorry for the delayed response. 
The `resType` as is here is `!fir.boxproc<(!fir.ref<i32>) -> !fir.boxproc<(!fir.ref<i32>) -> f32>>` for a function that returns a procedure pointer that has an interface of `real foo(integer)`
The test case I used for generating the code is
```
  interface
    real function func(x)
      integer :: x
    end
  end interface
  real :: res
  procedure(func), pointer :: arg0
  arg0=>reffunc(5)
contains
  function reffunc(arg1) result(pp)
    integer :: arg1
    procedure(func), pointer :: pp
    pp => func
  end
end
```

The FIR looks like:
```
...
%3:2 = hlfir.declare %0 {fortran_attrs = #fir.var_attrs<pointer>, uniq_name = "_QFEarg0"} : (!fir.ref<!fir.boxproc<(!fir.ref<i32>) -> f32>>) -> (!fir.ref<!fir.boxproc<(!fir.ref<i32>) -> f32>>, !fir.ref<!fir.boxproc<(!fir.ref<i32>) -> f32>>)
...
%7 = fir.call @_QFPreffunc(%6#1) fastmath<contract> : (!fir.ref<i32>) -> !fir.boxproc<(!fir.ref<i32>) -> f32> fir.store %7 to %3#0 : !fir.ref<!fir.boxproc<(!fir.ref<i32>) -> f32>>
```

DanielCChen replied:
Actually, after further investigation, you are right. `procTy.getResult(0)` indeed gives the correct return type of the function that returns the procedure pointer.

jeanPerier reviewed (commented):
Thanks you for working on this, the general approach looks good to me. Minor comments inlined.

jeanPerier commented:
It seems to be legal to use procedure pointer result in ASSOCIATED, this currently crashes:

```
 interface
   function foo()
     procedure(), pointer :: foo
   end function
 end interface
 print *, associated(foo())
 end
```

So either the fir.boxproc results needs to be placed in a storage after the call so that code that expects to get pointers works, or the ASSOCIATED lowering should accept fir.boxproc as first argument too.

https://github.com/llvm/llvm-project/blob/a90347ecc7e32c25a3b3820723f8405369f3abdc/flang/lib/Optimizer/Builder/IntrinsicCall.cpp#L2180

DanielCChen commented:
> It seems to be legal to use procedure pointer result in ASSOCIATED, this currently crashes:
> 
> ```
>  interface
>    function foo()
>      procedure(), pointer :: foo
>    end function
>  end interface
>  print *, associated(foo())
>  end
> ```
> 
> So either the fir.boxproc results needs to be placed in a storage after the call so that code that expects to get pointers works, or the ASSOCIATED lowering should accept fir.boxproc as first argument too.
> 
> https://github.com/llvm/llvm-project/blob/a90347ecc7e32c25a3b3820723f8405369f3abdc/flang/lib/Optimizer/Builder/IntrinsicCall.cpp#L2180

Good point. With the current implementation, I think I can make `genAssociated` to take `fir.boxproc` as the type of `arg[0]` and skip the `fir::load`.

DanielCChen committed:
[Flang] Support ASSOCIATED to have the first argument as a function that returns a procedure pointer.

DanielCChen committed:
[Flang] Add LIT test for ASSOCIATED with procedure pointer function result.

jeanPerier committed:
[flang] Lower passing non assumed-rank/size to assumed-ranks (#79145)

Start implementing assumed-rank support as described in
https://github.com/llvm/llvm-project/blob/main/flang/docs/AssumedRank.md

This commit holds the minimal support for lowering calls to procedure
with assumed-rank arguments where the procedure implementation is done
in C.

The case for passing assumed-size to assumed-rank is left TODO since it
will be done a change in assumed-size lowering that is better done in
another patch.

Care is taken to set the lower bounds to zero when passing non allocatable no pointer as descriptor
to a BIND(C) procedure as required per 18.5.3 point 3. This was not done before while the requirements also applies to non assumed-rank descriptors. This change  required special attention with IGNORE_TKR(t) to avoid emitting invalid fir.rebox operations (the actual argument type must be used in this case as the output type).

Implementation of Fortran procedure with assumed-rank arguments is still
TODO.

DanielCChen committed:
[Flang] Check ptrTy is not null before referencing it.

DanielCChen committed:
[Flang] Fixing code format.

DanielCChen commented on flang/lib/Lower/ConvertCall.cpp:
```diff
@@ -202,6 +205,8 @@ fir::ExtendedValue Fortran::lower::genCallOpAndResult(
     llvm::SmallVector<mlir::Value> lengths;
     if (!caller.callerAllocateResult())
       return {};
+    if (isProcedurePointer)
```
This is to respond the above comment at line 162. 
This change is actually no longer needed as I don't `setSaveResult` for procedure pointer function result. Therefore, call to `caller.callerAllocateResult` is returned `false` for procedure pointer function result.

DanielCChen committed:
[Flang] Address review comments.

DanielCChen committed:
[Flang] Minor clean up.

DanielCChen committed:
[Flang] To address review comment to correct the resultType.

jeanPerier reviewed (approved):
Looks great, thanks for the updates!
